crates/subspace-networking/src/utils.rs

```diff
@@ -150,20 +168,30 @@ impl SemState {
 
     // Expands the max capacity by delta.
     // Returns true if any waiters need to be notified.
-    fn expand(&mut self, delta: usize) -> bool {
+    fn expand(&mut self, delta: usize) -> Result<bool, SemaphoreError> {
         let prev_is_full = self.is_full();
-        self.capacity += delta;
-
-        // Notify if we did a full -> available transition.
-        prev_is_full && !self.is_full()
+        if let Some(capacity) = self.capacity.checked_add(delta) {
+            self.capacity = capacity;
+            // Notify if we did a full -> available transition.
+            Ok(prev_is_full && !self.is_full())
+        } else {
+            Err(SemaphoreError::InvalidExpand {
+                capacity: self.capacity,
+                delta,
+            })
+        }
     }
 
     // Shrinks the max capacity by delta.
-    fn shrink(&mut self, delta: usize) {
-        if let Some(dec) = self.capacity.checked_sub(delta) {
-            self.capacity = dec;
+    fn shrink(&mut self, delta: usize) -> Result<(), SemaphoreError> {
+        if let Some(capacity) = self.capacity.checked_sub(delta) {
+            self.capacity = capacity;
+            Ok(())
         } else {
-            panic!("SemState::shrink(): invalid shrink, state = {self:?}");
+            Err(SemaphoreError::InvalidShrink {
```

Comment:
This looks good, thanks. 

On some more thought, think we should return a handle for the expand op (like `ExpandPermit`), and have the shrink API take the expand permit and consume it. Or just drop the ExpandPermit, which would not shrink it back.

This would need some changes on node runner side (to keep map of peer -> ExpandPermit), but it would have caught the issue where disconnect messages are more than the connect ones

This can be done later, so you are unblocked now. Feel free to assign the ticket to me for follow up clean ups.

Reply:
I'm not sure how ergonomic it would be though. We may not necessarily want to shrink/expand on the same amounts each time, so it would be awkward to use if we get permit on one amount and want to free another. Returning a vector of permits is even worse in that case.

Reply:
yeah this assumes expand/shrink happen by same amount(our use case anyway). If functionality is all we need, we should just take the expand delta during `new()` and remove the delta from the expand/shrink APIs, to make it clear. But this is up to discussion